File: app/models/ai_analysis.py
from sqlalchemy import Column, Integer, String, DateTime, Text, ForeignKey, JSON, Float
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.core.database import Base


class AIAnalysis(Base):
    __tablename__ = "ai_analyses"

    id = Column(Integer, primary_key=True, index=True)
    incident_id = Column(Integer, ForeignKey("incidents.id", ondelete="CASCADE"), nullable=False, unique=True)

    # AI model info
    model_used = Column(String(50), nullable=False)  # e.g., "gpt-4", "claude-3"
    prompt_tokens = Column(Integer, nullable=True)
    completion_tokens = Column(Integer, nullable=True)
    total_cost_usd = Column(Float, nullable=True)

    # Analysis results
    root_cause_hypothesis = Column(Text, nullable=False)
    confidence_score = Column(Float, nullable=True)  # 0.0 to 1.0

    # Structured suggestions
    debug_checklist = Column(JSON, nullable=False)  # ["Check DNS", "Verify SSL cert"]
    suggested_actions = Column(JSON, nullable=False)  # [{"action": "...", "priority": "high"}]
    related_error_patterns = Column(JSON, nullable=True)

    # Full AI response
    raw_response = Column(Text, nullable=True)  # Store full LLM output

    # Timing
    analyzed_at = Column(DateTime(timezone=True), server_default=func.now())
    analysis_duration_ms = Column(Integer, nullable=True)

    # Relationships
    incident = relationship("Incident", back_populates="ai_analysis")

File: fe/lib/features/incident/domain/entities/ai_analysis.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'ai_analysis.freezed.dart';

/// AI Analysis domain entity
/// Contains AI-generated root cause analysis for an incident
@freezed
class AiAnalysis with _$AiAnalysis {
  const factory AiAnalysis({
    required int id,
    required int incidentId,
    required String modelUsed,
    int? promptTokens,
    int? completionTokens,
    double? totalCostUsd,
    required String rootCauseHypothesis,
    double? confidenceScore,
    required List<String> debugChecklist,
    required List<Map<String, dynamic>> suggestedActions,
    List<String>? relatedErrorPatterns,
    String? rawResponse,
    required DateTime analyzedAt,
    required int analysisDurationMs,
  }) = _AiAnalysis;

  const AiAnalysis._();

  /// Total tokens used
  int get totalTokens => (promptTokens ?? 0) + (completionTokens ?? 0);

  /// Formatted cost
  String get formattedCost => totalCostUsd != null ? '\$${totalCostUsd!.toStringAsFixed(4)}' : 'N/A';

  /// Formatted confidence
  String get formattedConfidence => confidenceScore != null ? '${(confidenceScore! * 100).toStringAsFixed(1)}%' : 'N/A';
}
